wgpu-core/src/device/life.rs

```diff
@@ -498,297 +499,306 @@ impl<A: HalApi> LifetimeTracker<A> {
                         self.active
                             .iter_mut()
                             .find(|a| a.index == submit_index)
                             .map_or(&mut self.free_resources, |a| &mut a.last_resources)
                             .compute_pipes
-                            .push(res.raw);
+                            .push(res);
                     }
                 }
             }
         }
 
         if !self.suspected_resources.render_pipelines.is_empty() {
-            let (mut guard, _) = hub.render_pipelines.write(token);
             let mut trackers = trackers.lock();
 
-            for id in self.suspected_resources.render_pipelines.drain(..) {
+            for render_pipeline in self.suspected_resources.render_pipelines.drain(..) {
+                let id = render_pipeline.info.id();
                 if trackers.render_pipelines.remove_abandoned(id) {
-                    log::debug!("Render pipeline {:?} will be destroyed", id);
+                    log::info!("RenderPipeline {:?} is removed from trackers", id);
                     #[cfg(feature = "trace")]
-                    if let Some(t) = trace {
-                        t.lock().add(trace::Action::DestroyRenderPipeline(id.0));
+                    if let Some(ref mut t) = trace {
+                        t.add(trace::Action::DestroyRenderPipeline(id.0));
                     }
 
-                    if let Some(res) = hub.render_pipelines.unregister_locked(id.0, &mut *guard) {
-                        let submit_index = res.life_guard.life_count();
+                    if let Some(res) = hub.render_pipelines.unregister(id.0) {
+                        let submit_index = res.info.submission_index();
                         self.active
                             .iter_mut()
                             .find(|a| a.index == submit_index)
                             .map_or(&mut self.free_resources, |a| &mut a.last_resources)
                             .render_pipes
-                            .push(res.raw);
+                            .push(res);
                     }
                 }
             }
         }
 
         if !self.suspected_resources.pipeline_layouts.is_empty() {
-            let (mut guard, _) = hub.pipeline_layouts.write(token);
+            let mut pipeline_layouts_locked = hub.pipeline_layouts.write();
 
-            for Stored {
-                value: id,
-                ref_count,
-            } in self.suspected_resources.pipeline_layouts.drain(..)
-            {
+            for pipeline_layout in self.suspected_resources.pipeline_layouts.drain(..) {
+                let id = pipeline_layout.info.id();
                 //Note: this has to happen after all the suspected pipelines are destroyed
-                if ref_count.load() == 1 {
+                if pipeline_layouts_locked.is_unique(id.0).unwrap() {
                     log::debug!("Pipeline layout {:?} will be destroyed", id);
```

Comment:
Might as well promote these from `log::debug` to `log::info` as well. Same for bind group layout.

Reply:
Definitely! The should all be log info